File: packages/shop-web-angular/src/app/services/geo-location.ts
import { Injectable } from '@angular/core';
import { ILocation } from '@modules/server.common/interfaces/IGeoLocation';
import GeoLocation from '@modules/server.common/entities/GeoLocation';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Subscribable } from 'rxjs';
import { environment } from 'environments/environment';

interface Coords {
	longitude?: number;
	latitude?: number;
}

@Injectable()
export class GeoLocationService {
	private headers: HttpHeaders = new HttpHeaders({
		'Content-Type': 'application/json',
	});

	constructor(private http: HttpClient) {}

	getCurrentGeoLocation(): Promise<GeoLocation> {
		return new Promise(async (resolve, reject) => {
			try {
				const coords = await this.getCurrentCoords();
				const location: ILocation = {
					type: 'Point',
					coordinates: [coords.longitude, coords.latitude],
				};

				const currentGeolocation = new GeoLocation({
					_id: '',
					loc: location,
					countryId: null,
					city: null,
					streetAddress: null,
					house: null,
					_createdAt: '',
					_updatedAt: '',
				});
				resolve(currentGeolocation);
			} catch (error) {
				reject(error);
			}
		});
	}

	getCurrentCoords(): Promise<Coords> {
		return new Promise((resolve, reject) => {
			const defaultLat = environment.DEFAULT_LATITUDE;
			const defaultLng = environment.DEFAULT_LONGITUDE;

			if (defaultLat && defaultLng) {
				resolve(
					this.getCoordsObj({
						latitude: defaultLat,
						longitude: defaultLng,
					})
				);

				return;
			}

			navigator.geolocation.getCurrentPosition(
				(res) => {
					// If user is enable GPS on browser
					resolve(this.getCoordsObj(res.coords));
				},
				(err) => {
					// If user is denied GPS on browser
					this.getLocationByIP().subscribe((res) => {
						if (res) {
							resolve(this.getCoordsObj(res));
						} else {
							reject(err.message);
						}
					});
				}
			);
		});
	}

	private getLocationByIP(): Subscribable<Coords | null> {
		return this.http.get(
			environment.SERVICES_ENDPOINT + '/getLocationByIP',
			{ headers: this.headers }
		);
	}

	private getCoordsObj(coords: Coords) {
		return {
			longitude: coords.longitude,
			latitude: coords.latitude,
		};
	}
}

File: packages/shop-mobile-flutter/lib/models/geolocation.dart
class Location {
  final String type;
  final List<double> coordinates;

  const Location({
    required this.type,
    required this.coordinates,
  });
}

class GeoLocationCoordinates {
  final double lng;
  final double lat;

  const GeoLocationCoordinates({
    required this.lng,
    required this.lat,
  });
}

class GeoLocation {
  final String? typename;
  final String? id;
  final DateTime? createdAt;
  final DateTime? updatedAt;
  final int countryId;
  final String? countryName;
  final String city;
  final String streetAddress;
  final String house;
  final String postcode;
  final String? notes;
  final Location loc;
  final GeoLocationCoordinates? coordinates;

  const GeoLocation({
    this.typename,
    this.id,
    this.createdAt,
    this.updatedAt,
    required this.countryId,
    this.countryName,
    required this.city,
    required this.streetAddress,
    required this.house,
    required this.postcode,
    this.notes,
    required this.loc,
    this.coordinates,
  });

  factory GeoLocation.fromJson(Map<String, dynamic> json) {
    return GeoLocation(
      typename: json['__typename'],
      id: json['_id'],
      createdAt: json['_createdAt'],
      updatedAt: json['_updatedAt'],
      countryId: json['countryId'],
      countryName: json['countryName'],
      city: json['city'],
      streetAddress: json['streetAddress'],
      house: json['house'],
      postcode: json['postcode'],
      notes: json['notes'],
      loc: json['loc'],
      coordinates: json['coordinates'],
    );
  }
}
